hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/Formatter.java

```diff
@@ -0,0 +1,791 @@
+import java.text.DecimalFormat;
+import java.text.DecimalFormatSymbols;
+
+import java.time.DayOfWeek;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.time.format.TextStyle;
+import java.time.temporal.ChronoField;
+import java.time.temporal.IsoFields;
+import java.time.temporal.JulianFields;
+import java.time.temporal.Temporal;
+import java.time.temporal.TemporalField;
+import static java.time.temporal.WeekFields.ISO;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.BiFunction;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Implementation of <a href="https://www.postgresql.org/docs/14/functions-formatting.html">
+ * PostgreSQL <code>to_char()</code></a> with the following differences.
+ * <ol><li> {@code FM} and {@code TM} prefixes enable the fill and translation modes for date
+ *          patterns respectively, which are disabled by default.
+ *     <li> {@code V} pattern can be combined with a decimal point.
+ *     <li> {@code M} and {@code P} patterns are introduced as the anchored versions of {@code MI}
+ *          and {@code PL} patterns respectively.
+ *     <li> {@code PR} pattern is replaced with {@code B} and {@code BR} patterns. The former is
+ *          the anchored version of the latter.
+ *     <li> {@code L} pattern is replaced with {@code C} and {@code CR} patterns. The former is the
+ *          anchored version of the latter.
+ *     <li> Literals are anchored by default. To fix its position, a literal should be prepended
+ *          with an {@code F} pattern.
+ *     <li> Zero-padding and space-padding are completely orthogonal, which makes it possible to
+ *          have zero-padded fractions, which are aligned at the the decimal separator. However,
+ *          this necessitates to have a {@code '0'} at the end of the pattern if the PostgreSQL
+ *          convention is desired. </ol>
+ */
+@SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:DeclarationOrder",
+        "checkstyle:ExecutableStatementCount", "checkstyle:InnerAssignment", "checkstyle:MagicNumber",
+        "checkstyle:MethodLength", "checkstyle:MultipleVariableDeclarations", "checkstyle:NestedIfDepth",
+        "checkstyle:NPathComplexity"})
+public class Formatter implements Serializable {
+    private static final Pattern DATETIME_TEMPLATE = Pattern.compile(
+            "((?:FM|TM)*)(SSSSS?|HH(?:12|24)?|MI|[SMU]S|FF[1-6]|[AP](?:M|\\.M\\.)|[ap](?:m|\\.m\\.)|"
+            + "DA?Y|[Dd]a?y|I?DDD|DD|I?D|W|[WI]W|CC|J|Q|MON(?:TH)?|[Mm]on(?:th)?|MM|RM|rm|"
+            + "Y,YYY|[YI]Y{0,3}|BC|B\\.C\\.|bc|b\\.c\\.|AD|A\\.D\\.|ad|a\\.d\\.|"
+            + "TZ[HM]?|TZ|tz|OF)(TH|th)?|\"[^\"]*\"");
+    private static final Pattern NUMERIC_TEMPLATE = Pattern.compile(
+            "FM|[90]+|[,G.D]|BR?|SG?|MI?|PL?|CR?|F?\"[^\"]*\"|V9+|TH|th|EEEE|RN");
```

Comment:
We should use a reluctant qualifier for the literal.
```suggestion
            "FM|[90]+|[,G.D]|BR?|SG?|MI?|PL?|CR?|F?\"[^\"]*?\"|V9+|TH|th|EEEE|RN");
```
Without it, if there are two quoted literals in the pattern, they will all match as a single literal:
```
to_char(1, '"prefix"000"postfix");
```

Reply:
Negation class `[^"]` cannot match `"` no matter what its qualifier is.

Reply:
Hmm, true, though I tested it and iirc, I reproduced it, but I must have made a mistake... :)